src/network/requestQueue/index.js

```diff
@@ -0,0 +1,117 @@
+const SocketRequest = require('./socketRequest')
+
+module.exports = class RequestQueue {
+  /**
+   * @param {number} maxInFlightRequests
+   * @param {number} requestTimeout
+   * @param {string} clientId
+   * @param {string} broker
+   * @param {Object} logger
+   */
+  constructor({ maxInFlightRequests, requestTimeout, clientId, broker, logger }) {
+    this.maxInFlightRequests = maxInFlightRequests
+    this.requestTimeout = requestTimeout
+    this.clientId = clientId
+    this.broker = broker
+    this.logger = logger
+
+    this.inflight = new Map()
+    this.pending = []
+  }
+
+  createRequest({ entry, expectResponse, send: sendRequest }) {
+    const request = new SocketRequest({
+      entry,
+      expectResponse,
+      broker: this.broker,
+      requestTimeout: this.requestTimeout,
+      send: () => {
+        this.inflight.set(entry.correlationId, request)
```

Comment:
It's a little weird that the created SocketRequest could flag an inflight request without having been `push`ed on to the queue (meaning the request never gets removed). I'd either move this binding to within `push` or just remove this method from the public api, meaning we'd call `push` with `{ entry, expectResponse, send }`.

Reply:
Yeah, I wasn't happy with this. The new approach is better 866eb43